kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/Eq.kt

```diff
@@ -15,15 +17,32 @@ interface Eq<T> {
 /**
  * Locates the applicable [Eq] for the inputs, and invokes it, returning the error if any.
  */
-fun <T> eq(actual: T, expected: T): Throwable? = when {
-   actual is Map<*, *> && expected is Map<*, *> -> MapEq.equals(actual, expected)
-   actual is Throwable && expected is Throwable -> ThrowableEq.equals(actual, expected)
-   actual is Regex && expected is Regex -> RegexEq.equals(actual, expected)
-   actual is String && expected is String -> StringEq.equals(actual, expected)
-   actual is Number && expected is Number -> NumberEq.equals(actual, expected)
-   shouldShowDataClassDiff(actual, expected) -> DataClassEq.equals(actual as Any, expected as Any)
-   else -> DefaultEq.equals(actual as Any, expected as Any)
+fun <T: Any?> eq(actual: T, expected: T): Throwable? {
+   // if we have null and non null, usually that's a failure, but people can override equals to allow it
+   return when {
+      actual == null && expected == null -> null
```

Comment:
` actual == null && expected == null -> null` should be handled in `TypesEq`.equals